### PR TITLE
Support for hexadecimal digits in regular expressions on the GPU

### LINKED ISSUE #4486: [BUG] Regular expressions with hex digits not working as expected
**Describe the bug**

I am seeing cuDF fail to match some hex digits correctly. Here is one example. I do not yet know if this is related to https://github.com/NVIDIA/spark-rapids/issues/4409 or whether this is a separate issue.

```
javaPattern=\xA9, cudfPattern=\xA9, input='©', cpu=true, gpu=false
```

**Steps/Code to reproduce bug**

```scala
  test("compare CPU and GPU: hex") {
    val patterns = Seq(raw"\x61", raw"\xA9", raw"\xa9")
    val inputs = Seq("a", "b", "\u00A9")
    assertCpuGpuMatchesRegexpFind(patterns, inputs)
  }
```

**Expected behavior**
Test should pass.

**Environment details (please complete the following information)**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #4486.

This code enables partial support for hexadecimal digits in regular expressions on the GPU. A couple of limitations that currently will fall back to CPU:

* Hex digits greater >= `0x80` are not supported
* Hex digits in character classes, e.g. `[\x20]`

Also, this code will transpile the variable length hex digits format from Java's `Pattern` into the 2 digit format that is supported by CUDF.

For example `\x{007f}` will transpile to `\x7f`, and `\x{f}` will transpile to `\x0f`.

Signed-off-by: Navin Kumar <navink@nvidia.com>

### CODE REVIEW

**Review by jlowe — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 1]
```diff
@@ -523,15 +528,20 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
-      case RegexHexDigit(_) =>
-        // see https://github.com/NVIDIA/spark-rapids/issues/4486
-        throw new RegexUnsupportedException(
-          s"cuDF does not support hex digits consistently with Spark")
+      case RegexHexDigit(digits) =>
+        val codePoint = Integer.parseInt(digits, 16)
+        if (codePoint >= 128) {
+          // see https://github.com/NVIDIA/spark-rapids/issues/4866
+          throw new RegexUnsupportedException(
+            "cuDF does not support hex digits > 0x80")
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:542]:
```suggestion
            "cuDF does not support hex digits > 0x7F")
```

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala [hunk 2]
```diff
@@ -144,13 +144,23 @@ class RegularExpressionTranspilerSuite extends FunSuite with Arm {
...
 
   test("cuDF does not support octal digits 0o177 < n <= 0o377") {
+    // see https://github.com/NVIDIA/spark-rapids/issues/4746
     val patterns = Seq(raw"\0200", raw"\0377")
     patterns.foreach(pattern =>
       assertUnsupported(pattern, RegexFindMode,
         "cuDF does not support octal digits 0o177 < n <= 0o377"))
   }
 
+  test("cuDF does not support hex digits > 0x80") {
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:154]:
```suggestion
  test("cuDF does not support hex digits > 0x7F") {
```

**Review by andygrove — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by NVnavkumar** [comment]
build

**Pull request merged**